### PR TITLE
Edit participants data

### PR DESCRIPTION
**Pull request by gchinis**
Hey Lukasz

Here is the change about the admin been able to modify all participants info (including Tshirts). Christoph told us that at the moment there is no requirement that the participants can order more than one tshirts, so in order to simplify the implementation we drop the shirts table and inlined the tshirt info into the participants table. Consecutively, the service layer was updated so that the methods that store and retrieve participants can also handle the shirt data.

About tests, a ParticipantBuilder was introduced that create mock Participants that are well formed(with all their fieds etc...).

As part of this story the semantics of the participant change as follows:
- A participant can order 0 or 1 t shirts.
- The participants.tshirt.details always have some default values and in order to know if a tshirt was actually order you need to look into the amount variable if it is 0 or 1. In the database this info in stored in the shirtordered column).

Please review it before merging :)

Thanks,
Georgios, Lucas, Liza

### CODE REVIEW

File: domain/tshirt.js [hunk 1]
```diff
@@ -10,19 +10,26 @@ function invalidData(body) {
   return _.isUndefined(body.size) || _.isUndefined(body.model);
 }
 
+tshirt.NoTShirt = function () {
```
Comment (lplotni) [inline domain/tshirt.js]: Why **NoTShirt** and not **noTShirt**?

File: domain/tshirt.js [hunk 2]
```diff
@@ -10,19 +10,26 @@ function invalidData(body) {
...
+  return {
+    details: [{ size: "L", model: "Slim fit" }],
```
Comment (lplotni) [inline domain/tshirt.js]: And does it make sense, to provide details if we know that, no tshirt has been ordered?

File: migrations/20160119150420-combine-shirts-with-participants.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+var dbm = global.dbm || require('db-migrate');
+var type = dbm.dataType;
+
+exports.up = function(db, callback) {
+  db.dropTable('tshirts', callback);
+
+  db.runSql("CREATE TYPE shirtsize AS ENUM ('XS', 'S', 'M', 'L', 'XL');",callback);
+  db.runSql("CREATE TYPE shirtmodel AS ENUM ('Normal fit', 'Slim fit');",callback);
+
+  db.addColumn('participants', 'shirtsize', {type: 'shirtsize', notNull:'true', defaultValue: 'XS'}, callback);
```
Comment (lplotni) [inline migrations/20160119150420-combine-shirts-with-participants.js]: Why do we need/want default values for the tshirt details?

File: service/participants.js [hunk 4]
```diff
@@ -17,11 +17,24 @@ let service = {};
 
 service._nodemailer = nodemailer;
 
+function participantFromRow(row) {
```
Comment (lplotni) [inline service/participants.js]: How about making this a factory function on the domain/participant?

File: service/participants.js [hunk 5]
```diff
@@ -56,18 +69,44 @@ service.delete = function (participantid) {
   return deferred.promise;
 };
 
+// TO BE REMOVED
```
Comment (lplotni) [inline service/participants.js]: Why still there?

File: spec/specHelper.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
+/* jshint node: true */
+/* jshint esnext: true */
+/* global jasmine */
+'use strict';
+
+let ParticipantBuilder = function() {
```
Comment (lplotni) [inline spec/specHelper.js]: :+1:

File: domain/tshirt.js (hunk 1 above)
Comment (gchinis) [inline domain/tshirt.js]: I capitalized to indicate that this is the NullObject of Tshirt rather than a method of a tshirt. I am not sure what is the convention if js for that. what do you think?

File: service/participants.js (hunk 4 above)
Comment (gchinis) [inline service/participants.js]: that is also possible but then the model will have to know the schema of the table. I put it in that file to avoid having table knowledge in the model

File: service/participants.js (hunk 5 above)
Comment (gchinis) [inline service/participants.js]: good catch thanks

File: domain/tshirt.js (hunk 2 above)
Comment (gchinis) [inline domain/tshirt.js]: The motivation behind the default values is to simplify the views. In the admin edit participant view there are elements that display tshirt info, even when these elements will end up hidden from the user, in the server side jade tries to fill these values  with information from the participant.

File: domain/tshirt.js (hunk 1 above)
Comment (lplotni) [inline domain/tshirt.js]: I'm not a fan of null-objects therefore I would suggest using something like an Optional.. but in general I'm still not sold on the whole idea, of having tshirt data in the partcipants object, even if there was no tshirt ordered.

File: domain/tshirt.js (hunk 2 above)
Comment (lplotni) [inline domain/tshirt.js]: In that case, this is a concern of the route and not really the tshirt domain object itself.

File: spec/specHelper.js (hunk 6 above)
Comment (lplotni) [inline spec/specHelper.js]: Could you extract the whole participantBuilder stuff to another pull request, then I could merge this straight away.

**Comment by lplotni** [comment]
I think we need to approach the whole tshirt <-> participant topic in a different manner. Yes, currently we're not supporting ordering of multiple tshirts but nevertheless by having a separate tshirt table, we have a nicer model. My main problem are the default values, why should we have any tshirt data at all, if nothing has been ordered?
